### PR TITLE
BUG: Implement PeriodEngine to fix PeriodIndex truncate bug

### LINKED ISSUE #17717: .truncate() fails on PeriodIndexes with duplicate values
Truncating a series with a PeriodIndex with unique entries works fine:
```python
In [4]: pdi = pd.PeriodIndex([pd.Period('2017-09-01'),
   ...:                       pd.Period('2017-09-02'),
   ...:                       pd.Period('2017-09-03')])
   ...: ser = pd.Series([1,2,3], index=pdi)
   ...: ser
Out[4]:
2017-09-01    1
2017-09-02    2
2017-09-03    3
Freq: D, dtype: int64

In [5]: ser.truncate(after='2017-09-02')
Out[5]:
2017-09-01    1
2017-09-02    2
Freq: D, dtype: int64
```
But truncating a series with a PeriodIndex with non-unique entries does not
```python
In [6]: pdi2 = pd.PeriodIndex([pd.Period('2017-09-02'),
   ...:                        pd.Period('2017-09-02'),
   ...:                        pd.Period('2017-09-03')])
   ...: ser2 = pd.Series([1,2,3], index=pdi2)
   ...: ser2
   ...:
Out[6]:
2017-09-02    1
2017-09-02    2
2017-09-03    3
Freq: D, dtype: int64

In [7]: ser2.truncate(after='2017-09-02')
```
This fails with a `TypeError: Cannot compare type 'Period' with type 'int'`. 

#### Expected Output
```python
2017-09-02    1
2017-09-02    2
Freq: D, dtype: int64
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 32
OS: Windows
OS-release: 7
machine: x86
processor: x86 Family 6 Model 42 Stepping 7, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: 0.3.1
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #17717
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by MaximilianR** [comment]
Does this make sense though? 

Even if it fixes that issue - why should a `PeriodIndex` ever be equal to an `Int64Index`?

**Comment by Licht-T** [comment]
@MaximilianR We are already treating Period and int as same.
https://github.com/Licht-T/pandas/blob/a240a734cd33362fb6825fd2ce67163e82f8f9a4/pandas/_libs/period.pyx#L761

**Comment by jreback** [comment]
@Licht-T no this is *adding* a well-defined value

```
In [2]: pd.Period('2012-02', freq='M') + 2
Out[2]: Period('2012-04', 'M')
```

this is NOT the same thing. We do not treat Periods and integers the same.

**Comment by jreback** [comment]
@Licht-T if you want to revise to the issue ok

**Comment by Licht-T** [comment]
@MaximilianR @jreback Thanks for your comment. I am making another solution.

**Comment by Licht-T** [comment]
@jreback @MaximilianR Now revised. I implemented `PeriodEngine`.

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -267,6 +269,10 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
         data = period.extract_ordinals(data, freq)
         return cls._from_ordinals(data, name=name, freq=freq)
 
+    @cache_readonly
```
Comment (jreback) [inline pandas/core/indexes/period.py]: this might already be implemented by a superclass

File: pandas/tests/indexes/period/test_indexing.py [hunk 2]
```diff
@@ -310,3 +311,60 @@ def test_take_fill_value(self):
 
         with pytest.raises(IndexError):
             idx.take(np.array([1, -5]))
+
+    def test_truncate(self):
+        # GH 17717
+        idx1 = pd.PeriodIndex([
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py]: these should go with the Series tests for truncate.

File: pandas/tests/indexes/period/test_indexing.py [hunk 3]
```diff
@@ -310,3 +311,60 @@ def test_take_fill_value(self):
...
+            pd.Period('2017-09-02'),
+            pd.Period('2017-09-02'),
+            pd.Period('2017-09-03')
+        ])
+        series1 = pd.Series([1, 2, 3], index=idx1)
+        result1 = series1.truncate(after='2017-09-02')
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py]: I would like to have a few period index specific tests that exercise the engine. you can look at what tests we have for example for DatetimeIndex / TimedeltaIndex for the same.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (Licht-T) [reply pandas/core/indexes/period.py]: @jreback Yes. But it force convert to `int64`.
https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/index_class_helper.pxi.in#L54

Comment (Licht-T) [reply pandas/core/indexes/period.py]: @jreback Yes, but the 1st argument is not same.
https://github.com/Licht-T/pandas/blob/6255812c30978c6a55f1758927f9c300e35cc82b/pandas/core/indexes/base.py#L1558

**Comment by Licht-T** [comment]
@jreback I'll add tests for `PeriodEngine`.

**Comment by Licht-T** [comment]
@jreback Added some tests

**Comment by Licht-T** [comment]
Fixed conflicts.

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
pls add a note for 0.22

**Comment by Licht-T** [comment]
@jreback Rebased & added what's new note.

File: pandas/_libs/index.pyx [hunk 4]
```diff
@@ -491,6 +494,76 @@ cdef class TimedeltaEngine(DatetimeEngine):
...
+    cdef _get_index_values(self):
+        return self.vgetter()
+
+    cpdef _call_map_locations(self, values):
+        super(PeriodEngine, self)._call_map_locations(values.view('i8'))
+
+    def _call_monotonic(self, values):
+        return super(PeriodEngine, self)._call_monotonic(values.view('i8'))
+
+    cdef _maybe_get_bool_indexer(self, object val):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: is this duplicating some of the super code?

File: pandas/_libs/index.pyx [hunk 5]
```diff
@@ -491,6 +494,76 @@ cdef class TimedeltaEngine(DatetimeEngine):
...
+            raise KeyError(val)
+        if count == 1:
+            return last_true
+
+        return result
+
+    def get_indexer(self, values):
+        cdef ndarray[int64_t, ndim=1] ordinals
+
+        self._ensure_mapping_populated()
```
Comment (jreback) [inline pandas/_libs/index.pyx]: can you call super in most of these cases?

File: pandas/_libs/index.pyx [hunk 6]
```diff
@@ -491,6 +494,76 @@ cdef class TimedeltaEngine(DatetimeEngine):
     cdef _get_box_dtype(self):
         return 'm8[ns]'
 
+
+cdef class PeriodEngine(Int64Engine):
+
+    cdef _get_index_values(self):
+        return self.vgetter()
+
+    cpdef _call_map_locations(self, values):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: is this necessary to create another way of calling this? e.g. ``_call_map_locations``

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/period.py]: I c. ok, we should make this much more generic (IOW just pass in self and have the engine extract what it needs), but that's another issue.

File: pandas/tests/indexes/period/test_indexing.py [hunk 7]
```diff
@@ -310,3 +311,191 @@ def test_take_fill_value(self):
...
+    def test_get_loc(self):
+        # GH 17717
+        p0 = pd.Period('2017-09-01')
+        p1 = pd.Period('2017-09-02')
+        p2 = pd.Period('2017-09-03')
+
+        idx0 = pd.PeriodIndex([p0, p1, p2])
+        expected_idx1_p1 = 1
+        expected_idx1_p2 = 2
+
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py]: can you add some comments on these cases

File: pandas/tests/indexes/period/test_indexing.py [hunk 8]
```diff
@@ -310,3 +311,191 @@ def test_take_fill_value(self):
...
+    def test_contains(self):
+        # GH 17717
+        p0 = pd.Period('2017-09-01')
+        p1 = pd.Period('2017-09-02')
+        p2 = pd.Period('2017-09-03')
+        p3 = pd.Period('2017-09-04')
+
+        ps0 = [p0, p1, p2]
+        idx0 = pd.PeriodIndex(ps0)
+
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py]: cc @jorisvandenbossche @MaximilianR @shoyer these are similar to Interval discussions, though conceptually much simpler.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/index.pyx (hunk 4 above)
Comment (Licht-T) [reply pandas/_libs/index.pyx]: Almost same, but where getting the `values` is not same.
https://github.com/Licht-T/pandas/blob/master/pandas/_libs/index_class_helper.pxi.in#L69
https://github.com/pandas-dev/pandas/pull/17755/files#diff-495f440aad68721ad5ca1f35087fc450R520

Comment (Licht-T) [reply pandas/_libs/index.pyx]: @jreback Almost same, but where getting the `values` is not same.
https://github.com/Licht-T/pandas/blob/master/pandas/_libs/index_class_helper.pxi.in#L69
https://github.com/pandas-dev/pandas/pull/17755/files#diff-495f440aad68721ad5ca1f35087fc450R520

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (Licht-T) [reply pandas/_libs/index.pyx]: @jreback So you mean that we don't have to define `_call_map_locations` by `cpdef`?

File: pandas/tests/indexes/period/test_indexing.py (hunk 8 above)
Comment (jorisvandenbossche) [reply pandas/tests/indexes/period/test_indexing.py]: Yes, and similar to that discussion, I don't like the current (and tested on the lines below) behaviour.

I don't think that:

```
In [97]: p0 = pd.Period('2017-09-01')
    ...: p1 = pd.Period('2017-09-02')

In [98]: idx = pd.PeriodIndex([p0, p1])

In [99]: idx.contains("2017-09-01 09:00")
Out[99]: True
```

should be the behaviour of contains. But let's leave that discussion out of this PR, as I don't think the PR is making any changes to the behaviour of `__contains__` / `contains` ? (or does it?)

File: pandas/_libs/index.pyx (hunk 4 above)
Comment (jreback) [reply pandas/_libs/index.pyx]: so provide a helper function to avoid repeating all of this code again.

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (Licht-T) [reply pandas/_libs/index.pyx]: @jreback I defined `_call_map_locations` as `cdef`, but some tests fail.
Seems that `_ensure_mapping_populated` is `inline`.

File: pandas/tests/indexes/period/test_indexing.py (hunk 8 above)
Comment (Licht-T) [reply pandas/tests/indexes/period/test_indexing.py]: @jorisvandenbossche @jreback Yes. No changes in `__contains__` / `contains`.

**Comment by Licht-T** [comment]
@jreback Almost fixed, except this.
https://github.com/pandas-dev/pandas/pull/17755#discussion_r148865409

File: pandas/tests/indexes/period/test_indexing.py [hunk 9]
```diff
@@ -3,10 +3,11 @@
 import pytest
 
 import numpy as np
+from numpy import testing as ntm
```
Comment (jreback) [inline pandas/tests/indexes/period/test_indexing.py]: NO, we never use numpy.testing.! remove!

use ``tm.assert_numpy_array_equal``.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Licht-T** [comment]
@jreback Removed numpy test module.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Licht-T if you'd like to see what can be simplified in index.pyx from the type hierarchy and/or general refactoring would be welcome.

**Comment by jreback** [comment]
one more indexer issue
https://travis-ci.org/MacPython/pandas-wheels/jobs/297752871

**Pull request merged**